pkg/parser/parser_csv.go

```diff
@@ -3,75 +3,82 @@ package parser
 import (
 	"bytes"
 	"encoding/csv"
-	"errors"
-	"fmt"
 	"io"
 )
 
+type CsvRow struct {
+	values []string
+	err    error
+}
+
 type CsvParser struct {
 }
 
 func (x CsvParser) Convert(unprocessed []byte) (doc Document, err error) {
-	d := Document{
+	doc = Document{
 		Attributes: Attributes{
 			Columns: make([]string, 0, 0),
 			Values:  make(map[string][]string),
 		},
 	}
 
-	defer func() (Document, error) {
-		if r := recover(); r != nil {
-			return d, errors.New(fmt.Sprintf("%s", r))
-		}
-		return d, err
-	}()
-
-	rowChannel := getCSVRows(bytes.NewReader(unprocessed))
+	rowChannel := getCsvRows(bytes.NewReader(unprocessed))
 	headerRow := <-rowChannel
 
+	if headerRow.err != nil {
+		return doc, headerRow.err
+	}
+
 	a := Attributes{
-		Columns: headerRow,
+		Columns: headerRow.values,
 		Values:  make(map[string][]string),
 	}
+
 	numCols := len(a.Columns)
 	for _, columnName := range a.Columns {
 		a.Values[columnName] = make([]string, 0)
 	}
 
-	for row := range rowChannel {
-		if len(row) != numCols {
+	for csvRow := range rowChannel {
+		if csvRow.err != nil {
+			return doc, csvRow.err
+		}
+		if len(csvRow.values) != numCols {
 			continue
 		}
 
-		for j, val := range row {
+		for j, val := range csvRow.values {
 			columnName := a.Columns[j]
 			a.Values[columnName] = append(a.Values[columnName], val)
 		}
 	}
 
-	d.Attributes = a
-	return d, err
+	doc.Attributes = a
+
+	return doc, err
 }
 
 // mostly https://stackoverflow.com/questions/32027590/efficient-read-and-write-csv-in-go
-func getCSVRows(rc io.Reader) (ch chan []string) {
-	ch = make(chan []string, 10)
+func getCsvRows(rc io.Reader) chan CsvRow {
+	ch := make(chan CsvRow, 10)
 	go func() {
 		r := csv.NewReader(rc)
 		defer close(ch)
 		for {
 			rec, err := r.Read()
-			if err != nil {
-				if err == io.EOF {
-					break
-				}
-				panic(err)
+			if err != nil && err == io.EOF {
+				break
+			}
 
+			r := CsvRow{
+				values: rec,
+				err:    err,
 			}
-			ch <- rec
+
+			ch <- r
```

Comment:
messed up and put the `err != nil` check in the if above. will put the check after this so that the error gets sent through the channel and then processing stops